src/routes/user/password.ts

```diff
@@ -1,29 +1,45 @@
 import { RequestHandler } from 'express';
 import { ReasonPhrases } from 'http-status-codes';
 
-import { gqlSdk, hashPassword } from '@/utils';
+import { gqlSdk, hashPassword, getUserByTicket } from '@/utils';
 import { sendError } from '@/errors';
 import { Joi, password } from '@/validation';
 
 export const userPasswordSchema = Joi.object({
   newPassword: password.required(),
+  ticket: Joi.string(),
 }).meta({ className: 'UserPasswordSchema' });
 
 export const userPasswordHandler: RequestHandler<
   {},
   {},
-  { newPassword: string }
+  { newPassword: string; ticket?: string }
 > = async (req, res) => {
-  // check if user is logged in
-  if (!req.auth?.userId) {
-    return sendError(res, 'unauthenticated-user');
+  const { ticket } = req.body;
+
+  // get the user from the ticket, but if no ticket then return null
```

Comment:
Performance: when using a ticket, the user is fetched twice, and updated twice. That's 4 operations. We could improve this in reviewing how conditions are nested. Something like this draft
```js
let user
if (ticket) {
  user = await getByTicket(ticket)
} else {
  if (!req.auth?.userId) { return sendError('...') }
  user = (await gqlSdk.user({ id: userId })).user
}

if (!user) { return sendError('...') }
const { newPassword } = req.body;
const passwordHash = await hashPassword(newPassword);

await gqlSdk.updateUser({
    id: user.id,
    user: {
      passwordHash,
      ticket: ticket ? null : undefined // not sure on this one, but I think Hasura undestands `undefined` as "do not change the value"
    },
  });
  return res.send(ReasonPhrases.OK);
};
```

Reply:
we could even use one single mutation like "update user.passwordHash, user.ticket where valid ticket OR id" and send an error if no user has been found:

```ts
const { ticket } = req.body;
  const userId = req.auth?.userId;
  if (!userId && !ticket) {
    return sendError(res, 'unauthenticated-user');
  }

  const { updateUsers } = await gqlSdk.updateUserWhere({
    where: {
      _or: [
        {
          _and: [
            {
              ticket: {
                _eq: ticket,
              },
            },
            {
              ticketExpiresAt: {
                _gt: new Date(),
              },
            },
          ],
        },
        { id: { _eq: userId } },
      ],
    },
    user: {
      passwordHash: await hashPassword(req.body.newPassword),
      ticket: ticket ? null : undefined,
    },
  });

  if (!updateUsers?.affected_rows) {
    return sendError(res, 'invalid-user-ticket'); // Invalid user or verification ticket
  }

  return res.send(ReasonPhrases.OK);
```

Reply:
thanks a lot for the review @plmercereau !

I went with your first suggestion, as I believe the added error granularity will be welcomed. I also confirmed that Hasura does not update a field if "undefined" is passed in as its value.

Have a look at the latest commit status and let me know what you think. Apologies for the duplicative commit, I got out of sync with the remote repository and had to merge the remote changes after I had already locally committed my own edits.